File: src/lib/shared/result.ts
interface ClientResult<T = unknown> {
	errorMessage?: string;
	isSuccess: boolean;
	value?: T;
}

export class ServerResult<T = unknown> {
	constructor(
		public isSuccess = false,
		public value?: T,
		public errorMessage?: string
	) {}

	public getClientResult(): ClientResult<T> {
		return {
			isSuccess: this.isSuccess,
			errorMessage: this.errorMessage,
			value: this.value
		};
	}
	public static success<T = unknown>(value?: T): ServerResult<T> {
		const result = new ServerResult<T>();

		result.value = value;
		result.isSuccess = true;
		result.errorMessage = undefined;

		return result;
	}

	public static failure(errorMessage: string): ServerResult<never> {
		const result = new ServerResult<never>();

		result.errorMessage = errorMessage;
		result.isSuccess = false;
		result.value = undefined;
		return result;
	}
}

File: apps/backend/src/Forge.Api/Features/Shared/Application/Common/ApplicationResult/Result.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Forge.Api.Features.Shared.Application.Common.ApplicationResult
{

    public interface IFailureResult<TSelf> where TSelf : IFailureResult<TSelf>
    {
        static abstract TSelf Failure(string errorMessage, int statusCode = 400);
    }
    public class Result : IFailureResult<Result>
    {
        public bool IsSuccess { get; }
        public bool IsFailure => !IsSuccess;
        public string ErrorMessage { get; protected set; } = string.Empty;
        public int StatusCode { get; }

        protected Result(bool isSuccess, string errorMessage, int statusCode)
        {
            IsSuccess = isSuccess;
            ErrorMessage = errorMessage;
            StatusCode = statusCode;

        }

        public static Result Success(int statusCode = 200)
           => new(true, string.Empty, statusCode);

        public static Result Failure(string errorMessage, int statusCode = 400)
            => new(false, errorMessage, statusCode);
    }
    public class Result<T> : Result, IFailureResult<Result<T>>
    {
        public T? Value { get; }

        protected Result(bool isSuccess, T? value, string errorMessage, int statusCode)
                : base(isSuccess, errorMessage, statusCode)
        {
            Value = value;
        }

        public static Result<T> Success(T? value, int statusCode = 200) => new(true, value, string.Empty, statusCode);
        public new static Result<T> Failure(string errorMessage, int statusCode = 400)
                => new(false, default, errorMessage, statusCode);


    }
}
